src/Tasks.UnitTests/Copy_Tests.cs

```diff
@@ -1406,6 +1406,32 @@ public void CopyFileOnItself2(bool isUseHardLinks, bool isUseSymbolicLinks)
             }
         }
 
+        [Theory]
+        [MemberData(nameof(GetHardLinksSymLinks))]
+        public void CopyFileItselfNonExistentFile(bool isUseHardLinks, bool isUseSymbolicLinks)
+        {
+            string temp = Path.GetTempPath();
+            string file = Path.Combine(temp, "FileThatDoesNotExist");
+
+            var engine = new MockEngine(_testOutputHelper);
+            var t = new Copy
+            {
+                RetryDelayMilliseconds = 1, // speed up tests!
+                BuildEngine = engine,
+                SourceFiles = new ITaskItem[] { new TaskItem(file) },
+                DestinationFiles = new ITaskItem[] { new TaskItem(file) },
+                SkipUnchangedFiles = false,
+                UseHardlinksIfPossible = isUseHardLinks,
+                UseSymboliclinksIfPossible = isUseSymbolicLinks,
+            };
+
+            bool success = t.Execute();
+
+            Assert.False(success);
+
+            ((MockEngine)t.BuildEngine).AssertLogDoesntContain("MSB3026"); // Didn't do retries, nothing to do
```

Comment:
We should probably as well assert on presence of 'MSB3030' ("Could not copy the file "{0}" because it was not found.")

Reply:
would a hypothetical `AssertLogDoesntContainAnyMSBCodes()` be useful

Reply:
`AssertLogDoesntContainAnyMSBCodes(List<string> except = null)` or `AssertLogContainsOnlyMSBCodes(List<string> allowedCodes = null)`

Reply:
Isn't part of the point of this that MSB3030 _should_ show up?

Reply:
Yes - I meant we should assert that the code *is* present.